---
title: "Making Figures"
author: "Tri Tran"
date: "April 27, 2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

##Set up packages needed

```{r}
library(dada2); packageVersion("dada2")
library(reshape2)
library(phyloseq)
library(ggplot2); packageVersion("ggplot2")
library(RColorBrewer)
library(data.table)
library(decontam)
library(magrittr)
library(DESeq2)
library(ggbiplot)
library(dplyr)
library(Rmisc)
library(broom)
library(VennDiagram)
library(vegan)
library(gplots)
source("Z:/data/TriTran/hr02192_AIP-root-microbiome_processed/DenefLab-MicrobeMiseq/R/miseqR.R") #will need to download from Denef Lab website.
# here: http://deneflab.github.io/MicrobeMiseq/ and unzip into code folder. There is also a copy of this code in the Lab 16S Processing Code folder. 
```

## Loading OTU processed data
 
```{r}
immun16.trim2 <- readRDS("immun16.trim2.rds")
immun16.trim2
```

```{r}
immun16.trim2 <- immun16.trim2 %>%
  subset_samples(
      SampleID != "EWV4" &
      SampleID != "EACD1" &
      SampleID != "RACD3" & 
      SampleID != "EPearson2" &
      SampleID != "RPearson1" &
      SampleID != "ECMII3" &
      SampleID != "EMM2" &
      SampleID != "EUC82B5" &
      SampleID != "RMM3" &
      SampleID != "RUC82B1" &
      SampleID != "E7132" &
      SampleID != "E8015"
  )
immun16.trim2
sum(otu_table(immun16.trim2)) #final total of seqs
mean(rowSums(otu_table(immun16.trim2))) #average number of seqs

sample_sum_df2 <- data.frame(sum = sample_sums(immun16.trim2))

# Histogram of sample read counts
ggplot(sample_sum_df2, aes(x = sum)) + 
  geom_histogram(color = "black", fill = "indianred", binwidth = 1000) +
  ggtitle("Distribution of sample sequencing depth") + 
  xlab("Read counts") +
  theme(axis.title.y = element_blank())

smin <- min(sample_sums(immun16.trim2))
smean <- mean(sample_sums(immun16.trim2))
smax <- max(sample_sums(immun16.trim2))

smin
smean
smax

```


## Overall Beta Diversity of All Samples Together separated by Compartment (if doing both endosphere and rhizosphere samples)

```{r}
# Scale reads to even depth 
immun16_scale <- immun16.trim2 %>%
  scale_reads(round = "round") 

# Ordinate
immun16_pcoa <- ordinate(
  physeq = immun16_scale, 
  method = "PCoA", 
  distance = "bray"
)
colourCount = 8
getPalette = colorRampPalette(brewer.pal(8, "Dark2"))
# Plot 
plot_ordination(
  physeq = immun16_scale,
  ordination = immun16_pcoa,
  color = "Compartment",
  axes = 1:2,
  title = "Bray-Curtis Dissimilarity PCoA") + 
  scale_color_manual(values = getPalette(colourCount)) +
  geom_point(aes(color = Compartment), alpha = 0.7, size = 5) +
  geom_point(colour = "grey90", size = 1.5) +
  theme(axis.text = element_text(size=12), legend.text = element_text(size=12)) +
  coord_fixed(ratio = 0.8)
```


Permanova results of overall Bray-Curtis Dissimilarity by compartment
```{r}
#Permanova
set.seed(1)

# Calculate bray curtis distance matrix
all.bray <- phyloseq::distance(immun16_scale, method = "bray")

# make a data frame from the sample_data
sampledf <- data.frame(sample_data(immun16_scale))

# Adonis test
adonis(all.bray ~ Compartment, data = sampledf)

# Homogeneity of dispersion test
beta <- betadisper(all.bray, sampledf$Compartment)
permutest(beta)
```

Alpha Diversity. Code from Denef lab tutorial.
```{r}
min_lib <- min(sample_sums(immun16.trim2))

# Initialize matrices to store richness and evenness estimates
nsamp = nsamples(immun16.trim2)
trials = 100

richness <- matrix(nrow = nsamp, ncol = trials)
row.names(richness) <- sample_names(immun16.trim2)

shannon <- matrix(nrow = nsamp, ncol = trials)
row.names(shannon) <- sample_names(immun16.trim2)


for (i in 1:100) {
  # Subsample
  r <- rarefy_even_depth(immun16.trim2, sample.size = min_lib, verbose = FALSE, replace = T)
  
  # Calculate richness
  rich <- as.numeric(as.matrix(estimate_richness(r, measures = "Observed")))
  richness[ ,i] <- rich
  
  # calculate shannon diversity
  shan <- as.numeric(as.matrix(estimate_richness(r, measures = "Shannon")))
  shannon[ ,i] <- shan
}

# Create a new dataframe to hold the means and standard deviations of richness estimates
SampleID <- row.names(richness)
mean <- apply(richness, 1, mean)
sd <- apply(richness, 1, sd)
measure <- rep("Richness", nsamp)
rich_stats <- data.frame(SampleID, mean, sd, measure)

# Create a new dataframe to hold the means and standard deviations of shannon diversity estimates
SampleID <- row.names(shannon)
mean <- apply(shannon, 1, mean)
sd <- apply(shannon, 1, sd)
measure <- rep("Shannon Diversity", nsamp)
shan_stats <- data.frame(SampleID, mean, sd, measure)

#combine richness and evenness into one dataframe
alpha <- rbind(rich_stats, shan_stats)

#adds sample data
s <- data.frame(sample_data(immun16.trim2))
alphadiv <- merge(alpha, s, by = "SampleID") 
alphadiv.hm <- subset(alphadiv, Study == "HM")
#alphadiv <- order(alphadiv, by="measure")
#setorder(alphadiv.hm, measure, Gen_com)
alphadiv.hm$Gen_com <- factor(alphadiv.hm$Gen_com, levels = c("Bulk_Bulk", "E_ACD", "E_UC82B", "R_ACD", "R_UC82B",
                                                        "E_Nr", "E_Pearson", "R_Nr", "R_Pearson", "E_def1", "E_CMII", "R_def1", "R_CMII",
                                                        "E_NahG", "E_MM", "R_NahG", "R_MM"))
```

## Plot Alpha Diversity

```{r}
colourCount = 8
getPalette = colorRampPalette(brewer.pal(8, "Dark2"))

#plot alpha diversity
ggplot(alphadiv.hm, aes(x = Gen_com, y = mean, color = Compartment, group = Gen_com)) +
  geom_boxplot() +
  #facet_wrap(~Mutant.Group, nrow = 2, scales = "free_x") +
  facet_grid(measure ~ Mutant.Group, scales = "free", space = "free_x") +
  geom_point(aes(fill = Compartment), size = 1, shape = 21, position = position_jitterdodge()) +
  scale_color_manual(values = getPalette(colourCount)) +
  scale_fill_manual(values = getPalette(colourCount)) +
  theme(axis.text.x = element_text(angle = 90), legend.position ="bottom")
```

#T-test for alpha diversity measures

```{r}
dat <- melt(alphadiv.hm)
dat <- subset(dat, variable == "mean" & Genotype !=  "Bulk" & measure == "Shannon Diversity")
shannon_result <- dat %>% group_by(Mutant.Group) %>% group_by(Compartment, add=T) %>% do(tidy(t.test(value~Genotype, data=.)))
shannon_result

```


# Scale reads to even depth

```{r}
hm.e <- immun16.trim2 %>%
  subset_samples(
    Study == "HM" &
    Compartment != "Bulk" &
      Compartment != "R")

hm_scale <- hm.e %>% scale_reads(round = "round")

# Ordinate
hm_pcoa <- ordinate(
  physeq = hm_scale, 
  method = "NMDS",
  trymax=100,
  distance = "bray")

#set genotype order
sample_data(hm_scale)[[3]] <- factor(sample_data(hm_scale)[[3]], levels = c("ACD", "UC82B", "Nr", "Pearson", "def1", "CMII", "NahG", "MM"))

#Plot
plot_ordination(
  physeq = hm_scale,
  ordination = hm_pcoa,
  color = "Mutant.Group",
  shape = "Defense",
  axes =c(1,2)
  ) +
   scale_color_manual(values = getPalette(colourCount)) +
  scale_shape_manual(values = c(15, 17)) +
  geom_point(aes(color = Mutant.Group), alpha = 0.7, size = 6) +
  coord_fixed(ratio = 0.95)
```

#Permanova

```{r}
# Calculate bray curtis distance matrix
all.bray <- phyloseq::distance(hm_scale, method = "bray")

# make a data frame from the sample_data
sampledf <- data.frame(sample_data(hm_scale))

# Adonis test
adonis(all.bray ~ Mutant.Group + Defense + Mutant.Group:Defense, data = sampledf)
```


#Differential Abundance Testing for Rhizosphere to endosphere enrichment

```{r}
#remove bulk
all.nobulk <- subset_samples(immun16.trim2, Compartment != "Bulk")

#convert to deseq object
diagdds = phyloseq_to_deseq2(all.nobulk, ~ Gen_com)

#calculate geometric means - alternate way that ignores NAs and 0s from Paul McMurdie phyloseq tutorial. 
#https://github.com/joey711/phyloseq/blob/master/vignettes/phyloseq-mixture-models.Rmd

gm_mean = function(x, na.rm=TRUE){
  exp(sum(log(x[x > 0]), na.rm=na.rm) / length(x))
}
geoMeans = apply(counts(diagdds), 1, gm_mean)
diagdds = estimateSizeFactors(diagdds, geoMeans = geoMeans)


diagdds.otu = DESeq(diagdds, fitType="local") #fit for OTU level

```

```{r}
##OTU level for ACD/UC82B rhizosphere to endosphere enrichment

RtoE_ACD <- results(diagdds.otu,
                    contrast=c("Gen_com", "E_ACD", "R_ACD"))
RtoE_ACD = RtoE_ACD[order(RtoE_ACD$padj, na.last=NA), ]
alpha = 0.05
sigtabRtoE_ACD = RtoE_ACD[(RtoE_ACD$padj < alpha), ]
sigtabRtoE_ACD = cbind(as(sigtabRtoE_ACD, "data.frame"), as(tax_table(all.nobulk)[rownames(sigtabRtoE_ACD), ], "matrix"))

RtoE_UC82B <- results(diagdds.otu,
                      contrast=c("Gen_com", "E_UC82B", "R_UC82B"))
RtoE_UC82B = RtoE_UC82B[order(RtoE_UC82B$padj, na.last=NA), ]
sigtabRtoE_UC82B = RtoE_UC82B[(RtoE_UC82B$padj < alpha), ]
sigtabRtoE_UC82B = cbind(as(sigtabRtoE_UC82B, "data.frame"), as(tax_table(all.nobulk)[rownames(sigtabRtoE_UC82B), ], "matrix"))
```

```{r}
#OTU level for NahG/MM endosphere
RtoE_NahG <- results(diagdds.otu,
                     contrast=c("Gen_com", "E_NahG", "R_NahG"))
RtoE_NahG = RtoE_NahG[order(RtoE_NahG$padj, na.last=NA), ]
sigtabRtoE_NahG = RtoE_NahG[(RtoE_NahG$padj < alpha), ]
sigtabRtoE_NahG = cbind(as(sigtabRtoE_NahG, "data.frame"), as(tax_table(all.nobulk)[rownames(sigtabRtoE_NahG), ], "matrix"))

RtoE_MM <- results(diagdds.otu,
                   contrast=c("Gen_com", "E_MM", "R_MM"))
RtoE_MM = RtoE_MM[order(RtoE_MM$padj, na.last=NA), ]
sigtabRtoE_MM = RtoE_MM[(RtoE_MM$padj < alpha), ]
sigtabRtoE_MM = cbind(as(sigtabRtoE_MM, "data.frame"), as(tax_table(all.nobulk)[rownames(sigtabRtoE_MM), ], "matrix"))
```


```{r}
#Subset to just enriched or depleted, then pull out rownames
RtoE_ACD.up <- subset(sigtabRtoE_ACD, sigtabRtoE_ACD$log2FoldChange > 0)
RtoE_ACD.up.list <-rownames(RtoE_ACD.up)
length(RtoE_ACD.up.list)

RtoE_ACD.down <- subset(sigtabRtoE_ACD, sigtabRtoE_ACD$log2FoldChange < 0)
RtoE_ACD.down.list <-rownames(RtoE_ACD.down)
length(RtoE_ACD.down.list)

RtoE_UC82B.up <- subset(sigtabRtoE_UC82B, sigtabRtoE_UC82B$log2FoldChange > 0)
RtoE_UC82B.up.list <-rownames(RtoE_UC82B.up)
length(RtoE_UC82B.up.list)

RtoE_UC82B.down <- subset(sigtabRtoE_UC82B, sigtabRtoE_UC82B$log2FoldChange < 0)
RtoE_UC82B.down.list <-rownames(RtoE_UC82B.down)
length(RtoE_UC82B.down.list)

RtoE_NahG.up <- subset(sigtabRtoE_NahG, sigtabRtoE_NahG$log2FoldChange > 0)
RtoE_NahG.up.list <-rownames(RtoE_NahG.up)
length(RtoE_NahG.up.list)

RtoE_NahG.down <- subset(sigtabRtoE_NahG, sigtabRtoE_NahG$log2FoldChange < 0)
RtoE_NahG.down.list <-rownames(RtoE_NahG.down)
length(RtoE_NahG.down.list)

RtoE_MM.up <- subset(sigtabRtoE_MM, sigtabRtoE_MM$log2FoldChange > 0)
RtoE_MM.up.list <-rownames(RtoE_MM.up)
length(RtoE_MM.up.list)

RtoE_MM.down <- subset(sigtabRtoE_MM, sigtabRtoE_MM$log2FoldChange < 0)
RtoE_MM.down.list <-rownames(RtoE_MM.down)
length(RtoE_MM.down.list)
```

## Venn Diagrams of relative abundance of enriched and depleted microbes in ACD and UC82B

```{r}
one <- RtoE_ACD.down.list
two <- RtoE_UC82B.down.list
three <- RtoE_ACD.up.list
four <- RtoE_UC82B.up.list

venn.plot <- draw.quad.venn(
  area1 = length(one),
  area2 = length(two),
  area3 = length(three),
  area4 = length(four),
  n12 = length(intersect(one, two)),
  n13 = length(intersect(one, three)),
  n23 = length(intersect(two, three)),
  n14 = length(intersect(one, four)),
  n24 = length(intersect(two, four)),
  n34 = length(intersect(three, four)),
  n123 = length(Reduce(intersect, list(one, two, three))),
  n124 = length(Reduce(intersect, list(one, two, four))),
  n134 = length(Reduce(intersect, list(one, three, four))),
  n234 = length(Reduce(intersect, list(two, three, four))),
  n1234 = length(Reduce(intersect, list(one, two, three, four))),
  category = c("ACD Down", "UC82B Down", 
               "ACD Up ", "UC82B Up"),
  fill = c("#1B9E77", "#D95F02", "#7570B3", "#E7298A"),
  lwd = rep(.6, 4), 
  lty = "solid", 
  col = "black",
  fontfamily ="sans",
  cat.fontfamily = "sans",
  cex = 1,
  cat.cex = .7,
  margin=0.15,
  euler.d = FALSE,
  scaled =  FALSE)
```


## Venn Diagrams of relative abundance of enriched and depleted microbes in NahG and MM

```{r}
one <- RtoE_NahG.down.list
two <- RtoE_MM.down.list
three <- RtoE_NahG.up.list
four <- RtoE_MM.up.list
venn.plot <- draw.quad.venn(
  area1 = length(one),
  area2 = length(two),
  area3 = length(three),
  area4 = length(four),
  n12 = length(intersect(one, two)),
  n13 = length(intersect(one, three)),
  n23 = length(intersect(two, three)),
  n14 = length(intersect(one, four)),
  n24 = length(intersect(two, four)),
  n34 = length(intersect(three, four)),
  n123 = length(Reduce(intersect, list(one, two, three))),
  n124 = length(Reduce(intersect, list(one, two, four))),
  n134 = length(Reduce(intersect, list(one, three, four))),
  n234 = length(Reduce(intersect, list(two, three, four))),
  n1234 = length(Reduce(intersect, list(one, two, three, four))),
  category = c("NahG Down", "MM Down", 
               "NahG Up ", "MM Up"),
  fill = c("#1B9E77", "#D95F02", "#7570B3", "#E7298A"),
  lwd = rep(.6, 4), 
  lty = "solid", 
  col = "black",
  fontfamily ="sans",
  cat.fontfamily = "sans",
  cex = 1,
  cat.cex = .7,
  margin=0.15,
  euler.d = FALSE,
  scaled =  FALSE
)
```

##Venn Diagrams of relative abundance of depleted microbes in ACD and NahG

```{r}
one <- RtoE_ACD.down.list
two <- RtoE_NahG.down.list

venn.plot <- draw.pairwise.venn(
  area1 = length(one),
  area2 = length(two),
  cross.area = length(intersect(one, two)),
  category = c("ACD Down", "NahG Down"),
  fill = c("#1B9E77", "#D95F02") ,
  lwd = rep(.6, 2), 
  lty = "solid", 
  col = "black",
  fontfamily ="sans",
  cat.fontfamily = "sans",
  cex = 1,
  cat.cex = .6,
  margin = 0.1, 
  euler.d = FALSE,
  scaled = FALSE
)
```

## Venn Diagrams of relative abundance of enirched microbes in ACD and NahG

```{r}
one <- RtoE_ACD.up.list
two <- RtoE_NahG.up.list

#twoway
venn.plot <- draw.pairwise.venn(
  area1 = length(one),
  area2 = length(two),
  cross.area = length(intersect(one, two)),
  category = c("ACD Up", "NahG Up"),
  fill = c("#1B9E77", "#D95F02") ,
  lwd = rep(.6, 2), 
  lty = "solid", 
  col = "black",
  fontfamily ="sans",
  cat.fontfamily = "sans",
  cex = 1,
  cat.cex = .6,
  margin = 0.1, 
  euler.d = FALSE,
  scaled = FALSE
)
```

